### DISCUSSION TITLE
Tradeoffs in SVD performance methods

### DISCUSSION THREAD

**Discussion by PabloAndresCQ** [NVIDIA/cuQuantum#97] [Q&A] [answered]
Hi dev team,

I have two related questions on SVD performance. EDIT: To clarify, I am using `tensor.decompose`.

**Context**. I am developing some approximate TN state evolution methods in [pytket-cutensornet](https://github.com/CQCL/pytket-cutensornet). As such, I need to use SVD decompositions quite often and I have been looking at the different SVD algorithms you provide to see which ones are best for my use case (see tables below). My conclusion is:

- Since `svdp` is the fastest, use it whenever the workspace required for decomposition would fit in memory.
- Otherwise, default to `svd`, since it's the most memory efficient.

![image](https://github.com/NVIDIA/cuQuantum/assets/104848389/7fe3d815-e59c-4812-a3de-308e70edcf3f)
 (top right of each table shows the memory footprint of the tensor to be decomposed)

**Question 1**. By using `logging.basicConfig(level=logging.DEBUG` I could see how much memory was allocated for the decomposition's workspace. Is there a way to obtain this value from the python API _without_ carrying out the decomposition?

**Question 2**. Even the default `svd` algorithm uses a considerable amount of memory (roughly x3 the size of the tensor to be decomposed). Are there any other SVD algorithms you could provide that minimise the amount of device memory required (at the expense of increasing runtime)? Potentially using the CPU and its RAM, if that would help?

Cheers!
Pablo

**Comment by DmitryLyakh** [comment]
For Question 2, you may try to use CUDA managed memory (combined CPU and GPU RAM) via cudaMallocManaged to see if it works for you and if the performance overhead is acceptable (it will depend on your system).

**Reply by PabloAndresCQ** [reply]
Thanks! Sounds interesting. Currently I am calling the SVD algorithms via the `tensor.decompose` API of cuTensorNet:
```
U, S, R = tensor.decompose("ik->ij,jk", a, method={'algorithm': 'gesvd'})
```
which manages workspace memory on its own. Would I be able to ask it to use unified memory by somehow passing `cudaMallocManaged` to the [allocator option](https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.cutensornet.tensor.DecompositionOptions.html#cuquantum.cutensornet.tensor.DecompositionOptions.allocator)? Or would I need to use [tensor_svd](https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.cutensornet.tensor_svd.html?highlight=tensor_svd#cuquantum.cutensornet.tensor_svd) and provide it an appropriate `workspace`?

**Reply by yangcal** [reply, marked as answer]
Hello Pablo,

Thanks for providing your benchmark data. A few points here:

> Since svdp is the fastest, use it whenever the workspace required for decomposition would fit in memory.

In our experiments, we did observe `gesvdp` being faster than `gesvd` in many cases. But there are some caveats in accuracy, please refer to cusolver's documentation [here](https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdnxgesvdp). When using `gesvdp` in `cuquantum`, we encourage you to check `SVDInfo.gesvdp_err_sigma` obtained via `tensor.decompose(..., return_info=True)` to make sure everything is well converged.

> Question 1. Is there a way to obtain this value from the python API without carrying out the decomposition?
`tensor.decompose` is designed as a **one-shot** API that includes meta data pre-processing, workspace calculation and execution. If you want to have some self-adjust svd implementation, you would need to hack into the workspace allocation part involving this [block](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/python/cuquantum/cutensornet/tensor.py#L275C65-L275C65) 

> Question 2. Are there any other SVD algorithms you could provide that minimise the amount of device memory required (at the expense of increasing runtime)?

`gesvd/gesvdj/gesvdp/gesvdp` are all the SVD algorithms that we currently support (`gesvdr` actually uses host memory as well, but not for saving device mem usage). However, like @DmitryLyakh mentioned above, you can use `cudaMallocManaged` to oversubscribe device mem. This can be done via the allocator option or simply via `cp.cuda.set_allocator(cp.cuda.malloc_managed)` before your workflow. Note that there may be performance drop depending on the amount of gpu oversubscription. We encourage you to do some benchmark based on your system and problem of interest. For instance, you may do `tmp = cp.empty(hold_size, dtype=cp.int8)` where you can adjust `hold_size` to modify the oversubscription for you fixed problem (assuming you already know how much workspace is needed for that problem). Such benchmark cases can give you an idea of how much potential slowdown you will be expecting.

Let us know if you have any more questions.

**Reply by PabloAndresCQ** [reply]
Thanks Yang! That's helpful, I'll be trying out these ideas in the coming weeks.

**Reply by leofang** [reply]
One minor improvement when using managed memory through CuPy is that you can guard it behind a mempool:
```py
cp.cuda.set_allocator(cp.cuda.MemoryPool(cp.cuda.malloc_managed).malloc)
```
so that you can reuse memory without frequent calls to `cudaMallocManaged`/`cudaFree`, which is costly.

See https://docs.cupy.dev/en/stable/user_guide/memory.html#changing-memory-pool for further detail.

**Reply by leofang** [reply]
> Is there a way to obtain this value from the python API _without_ carrying out the decomposition?

@PabloAndresCQ I would love to understand better this question. What would you do assuming we allow you to query the workspace size? If it's too large to fit in your budget, would it make any difference (since you can't run the decomposition anyway)?

Note that we do provide C & low-level Python binding for the workspace size query. But there's a gap between that and the high-level pythonic `tensor.decompose()` API, as you need to translate the problem into proper arguments to the former. If we understand your intention better, we might be able to offer better suggestion.

**Reply by PabloAndresCQ** [reply]
@leofang I was thinking of doing this kind of workflow:

1. Query the workspace required by the `gesvdp` method. If it fits in memory use it, since it's the fastest for my application.
2. Otherwise, query the workspace required by `gesvd`. If it fits, use it.
3. Otherwise, catch the OOM error before it happens and hadle it somehow. There might be multiple options: 
  a. free up GPU memory being used to store other tensors in the TN (e.g. truncating these, or temporarily moving them out of GPU mem);
  b. choose to use CUDA managed memory for this particular SVD decomposition, so that host RAM can be used;
  c. or simply inform the user of the problem and provide application-specific suggestions on what parameters to change to hopefully circumvent the OOM error. (Ofc, this latter one could be done by a standard try-except block, though).

Personally, I am willing to look into the low-level Python bindings as Yang suggested. But perhaps it'd be a nice feature if users could provide `tensor.decompose()` with an ordered list of preferred SVD methods so that cuTensorNet executes the first one that fits in memory (as in steps (1) and (2) above).

**Reply by leofang** [reply]
Thanks, Pablo, we'll discuss and get back to you. FWIW, the low-level API that I referred to is this:
https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.cutensornet.workspace_compute_svd_sizes.html
